minicluster/src/main/java/alluxio/master/MultiMasterEmbeddedJournalLocalAlluxioCluster.java

```diff
@@ -0,0 +1,310 @@
+   * @param index the index
+   * @return the worker process by index
+   */
+  public WorkerProcess getWorkerProcess(int index) {
+    return mWorkers.get(index);
+  }
+
+  /**
+   * @return index of leader master in {@link #mMasters}, or -1 if there is no leader temporarily
+   */
+  public int getLeaderIndex() {
+    for (int i = 0; i < mNumOfMasters; i++) {
+      if (mMasters.get(i).isServing()
+          && mMasters.get(i).getMasterProcess().mLeaderSelector.getState() == NodeState.PRIMARY) {
+        return i;
+      }
+    }
+    return -1;
+  }
+
+  /**
+   * @return the master addresses
+   */
+  public List<InetSocketAddress> getMasterAddresses() {
+    List<InetSocketAddress> addrs = new ArrayList<>();
+    for (int i = 0; i < mNumOfMasters; i++) {
+      addrs.add(mMasters.get(i).getAddress());
+    }
+    return addrs;
+  }
+
+  /**
+   * Iterates over the masters in the order of master creation, stops the first standby master.
+   *
+   * @return true if a standby master is successfully stopped, otherwise, false
+   */
+  public boolean stopStandby() {
+    for (int k = 0; k < mNumOfMasters; k++) {
+      if (!mMasters.get(k).isServing()) {
+        try {
+          LOG.info("master {} is a standby. stopping it...", k);
+          mMasters.get(k).stop();
+          LOG.info("master {} stopped.", k);
+        } catch (Exception e) {
+          LOG.error(e.getMessage(), e);
+          return false;
+        }
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Iterates over the masters in the order of master creation, stops the leader master.
+   *
+   * @return true if the leader master is successfully stopped, false otherwise
+   */
+  public boolean stopLeader() {
+    int leaderId = getLeaderIndex();
+    try {
+      LOG.info("master {} is the leader. stopping it...", leaderId);
+      getLocalAlluxioMasterByIndex(leaderId).stop();
+      LOG.info("master {} stopped.", leaderId);
+    } catch (Exception e) {
+      LOG.error(e.getMessage(), e);
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * Waits for the primary master to start until a timeout occurs.
+   *
+   * @param timeoutMs the number of milliseconds to wait before giving up and throwing an exception
+   */
+  public void waitForPrimaryMasterServing(int timeoutMs)
+      throws TimeoutException, InterruptedException {
+    CommonUtils.waitFor("the primary leader master to start",
+        () -> {
+          int leaderId = getLeaderIndex();
+          if (leaderId == -1) {
+            return false;
+          }
+          try {
+            getLocalAlluxioMasterByIndex(leaderId).getClient().listStatus(
+                new AlluxioURI("/"));
+            return true;
+          } catch (Exception e) {
+            return false;
+          }
+        },
+        WaitForOptions.defaults().setTimeoutMs(timeoutMs));
+  }
+
+  @Override
+  protected void startMasters() throws IOException {
+    for (int k = 0; k < mNumOfMasters; k++) {
```

Comment:
This method has a few caveats worth mentioning:

After you start a master, the config are all updated. So if you do `Configuration.get(PropertyKey.MASTER_RPC_PORT)` in a running thread, the return value may not be the correct one. It may be the port used by the next master in the line. It's fine to keep behavior as-is here, but let's add a comment for this tricky behavior. 

As said above, if internal to the master logic there's something resolving the master's RPC port, that logic will get a wrong port so that logic will definitely fail (master cannot start). Luckily, that logic described above does not exist. So your code here can work and give you a functioning cluster of multiple masters. Because the Configuration is a singleton in the process, there's nothing we can do.

Another caveat will be, it's safe to start masters one by one, instead of spawning multiple threads concurrently. That's due to the same problem mentioned above. Just one more reason to add a comment (so the next smart guy don't optimize this part and fall in a pit).